src/cargo/util/config.rs

```diff
@@ -675,14 +700,19 @@ fn walk_tree<F>(pwd: &Path, mut walk: F) -> CargoResult<()>
     where F: FnMut(File, &Path) -> CargoResult<()>
 {
     let mut current = pwd;
+    let mut stash: HashSet<ConfigFile> = HashSet::new();
 
     loop {
-        let possible = current.join(".cargo").join("config");
-        if fs::metadata(&possible).is_ok() {
-            let file = try!(File::open(&possible));
+        let possible = ConfigFile::new(current.join(".cargo").join("config"));
+        if possible.exist() && stash.get(&possible).is_none() {
```

Comment:
I think I can't do this. With `insert` I'm moving the value, so I can't use it inside the `if` block because it is out of scope, it is borrowed to the `HashSet`.